File: week-1/problem-set/prob8.py
"""Prob8 example.

This file belongs to the worked week 1 problem-set solutions. It shows two
different deletion ideas: constant-time deletion in an unsorted list, and lazy
deletion inside a sorted list.
"""


def delete_element(arr, i):
    """Delete one list element in constant time by swapping it with the last value.

    Args:
        arr (list): The list being processed by the algorithm.
        i (int): The i input used by this function.

    Returns:
        list: The list produced by the algorithm.
    """
    if not 0 <= i < len(arr):
        raise IndexError("Index out of range")

    # We lose the original order, but the removal itself is O(1).
    arr[i], arr[-1] = arr[-1], arr[i]
    arr.pop()
    return arr


class SortedArray:
    """Sorted array that marks deleted positions instead of removing them."""

    def __init__(self, arr):
        """Store the sorted values and a matching list of deletion markers.

        Args:
            arr (list): The list being processed by the algorithm.

        Returns:
            None: This function updates data or prints results and does not return a value.
        """
        self.arr = list(arr)
        self.deleted = [False] * len(arr)

    def _nearest_active_index(self, mid, left, right):
        """Return the closest undeleted index around the midpoint.

        Args:
            mid (int): The mid input used by this function.
            left (object): The left boundary or left-side data used by the algorithm.
            right (object): The right boundary or right-side data used by the algorithm.

        Returns:
            int: The integer result produced by the algorithm.
        """
        offset = 0
        while mid - offset >= left or mid + offset <= right:
            left_candidate = mid - offset
            right_candidate = mid + offset

            if left_candidate >= left and not self.deleted[left_candidate]:
                return left_candidate
            if right_candidate <= right and not self.deleted[right_candidate]:
                return right_candidate

            offset += 1

        return None

    def delete(self, i):
        """Mark the value at index i as deleted.

        Args:
            i (int): The i input used by this function.

        Returns:
            None: This function updates data or prints results and does not return a value.
        """
        if not 0 <= i < len(self.arr):
            raise IndexError("Index out of range")
        self.deleted[i] = True

    def search(self, x):
        """Search for x while skipping positions that were lazily deleted.

        Args:
            x (int): The x input used by this function.

        Returns:
            int: The integer result produced by the algorithm.
        """
        left = 0
        right = len(self.arr) - 1

        while left <= right:
            mid = (left + right) // 2
            active_mid = self._nearest_active_index(mid, left, right)

            if active_mid is None:
                return -1

            if self.arr[active_mid] < x:
                left = active_mid + 1
            elif self.arr[active_mid] > x:
                right = active_mid - 1
            else:
                return active_mid

        return -1

    def __len__(self):
        """Return the number of values that have not been deleted.

        Returns:
            int: The integer result produced by the algorithm.
        """
        return sum(1 for flag in self.deleted if not flag)

    def __getitem__(self, i):
        """Return the value at index i if it has not been deleted.

        Args:
            i (int): The i input used by this function.

        Returns:
            object: The value produced by this function.
        """
        if self.deleted[i]:
            raise IndexError("That position has been deleted")
        return self.arr[i]

    def __iter__(self):
        """Yield only the values that are still active.

        Returns:
            generator: A generator that yields values one at a time.
        """
        for index, value in enumerate(self.arr):
            if not self.deleted[index]:
                yield value


def main():
    """Run the small demonstration for this teaching example.

    Returns:
        None: This function updates data or prints results and does not return a value.
    """
    arr = [1, 2, 3, 4, 5]
    delete_index = 2

    print("Delete from an unsorted list:", delete_element(arr.copy(), delete_index))

    sorted_arr = SortedArray(arr)
    sorted_arr.delete(delete_index)
    print("Number of active values:", len(sorted_arr))
    print("Active values:", list(sorted_arr))
    print("Index of 4:", sorted_arr.search(4))


if __name__ == "__main__":
    main()

File: week-1/problem-set/prob8.cpp
/**
 * @file prob8.cpp
 * @brief C++ version of the deletion examples for unsorted and sorted arrays.
 */

#include <iostream>
#include <stdexcept>
#include <vector>

using namespace std;

/**
 * @brief Delete one element in constant time by swapping it with the last value.
 * @param values vector<int>& Unsorted list of integers that will be modified in place.
 * @param index size_t Position of the element to remove.
 * @return vector<int> Copy of the updated list after the deletion.
 */
vector<int> delete_element(vector<int>& values, size_t index) {
    if (index >= values.size()) {
        throw out_of_range("Index out of range");
    }

    swap(values[index], values.back());
    values.pop_back();
    return values;
}

class SortedArray {
public:
    /**
     * @brief Store sorted values and a matching deletion-marker array.
     * @param values vector<int> Sorted list of integers managed by the object.
     */
    explicit SortedArray(vector<int> values) : values_(move(values)), deleted_(values_.size(), false) {}

    /**
     * @brief Mark one position as deleted without physically removing it.
     * @param index size_t Position to mark as deleted.
     * @return void This method updates internal state and does not return a value.
     */
    void delete_index(size_t index) {
        if (index >= values_.size()) {
            throw out_of_range("Index out of range");
        }
        deleted_[index] = true;
    }

    /**
     * @brief Search for a target value while skipping lazily deleted positions.
     * @param target int Value to search for in the sorted array.
     * @return int Index of the target if it is still active, otherwise -1.
     */
    int search(int target) const {
        int left = 0;
        int right = static_cast<int>(values_.size()) - 1;

        while (left <= right) {
            int mid = (left + right) / 2;
            int active_mid = nearest_active_index(mid, left, right);
            if (active_mid == -1) {
                return -1;
            }
            if (values_[active_mid] < target) {
                left = active_mid + 1;
            } else if (values_[active_mid] > target) {
                right = active_mid - 1;
            } else {
                return active_mid;
            }
        }

        return -1;
    }

    /**
     * @brief Count how many positions are still active.
     * @return size_t Number of entries that have not been deleted.
     */
    size_t active_count() const {
        size_t count = 0;
        for (bool flag : deleted_) {
            if (!flag) {
                ++count;
            }
        }
        return count;
    }

    /**
     * @brief Return the active values as a new vector.
     * @return vector<int> Copy of the values whose deletion marker is false.
     */
    vector<int> active_values() const {
        vector<int> result;
        for (size_t index = 0; index < values_.size(); ++index) {
            if (!deleted_[index]) {
                result.push_back(values_[index]);
            }
        }
        return result;
    }

private:
    /**
     * @brief Find the nearest index around mid that has not been lazily deleted.
     * @param mid int Middle index of the current binary-search range.
     * @param left int Left boundary of the current binary-search range.
     * @param right int Right boundary of the current binary-search range.
     * @return int Closest active index, or -1 when no active position exists in the range.
     */
    int nearest_active_index(int mid, int left, int right) const {
        int offset = 0;
        while (mid - offset >= left || mid + offset <= right) {
            int left_candidate = mid - offset;
            int right_candidate = mid + offset;

            if (left_candidate >= left && !deleted_[left_candidate]) {
                return left_candidate;
            }
            if (right_candidate <= right && !deleted_[right_candidate]) {
                return right_candidate;
            }
            ++offset;
        }
        return -1;
    }

    vector<int> values_;
    vector<bool> deleted_;
};

/**
 * @brief Print a vector of integers in bracket form.
 * @param values const vector<int>& Vector to print. Each element has type int.
 * @return void This helper only prints data and does not return a value.
 */
void print_vector(const vector<int>& values) {
    cout << '[';
    for (size_t index = 0; index < values.size(); ++index) {
        cout << values[index];
        if (index + 1 != values.size()) {
            cout << ", ";
        }
    }
    cout << ']';
}

/**
 * @brief Run the deletion demonstrations.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    vector<int> values = {1, 2, 3, 4, 5};
    size_t delete_index_value = 2;

    vector<int> unsorted_copy = values;
    cout << "Delete from an unsorted list: ";
    print_vector(delete_element(unsorted_copy, delete_index_value));
    cout << '\n';

    SortedArray sorted_array(values);
    sorted_array.delete_index(delete_index_value);
    cout << "Number of active values: " << sorted_array.active_count() << '\n';
    cout << "Active values: ";
    print_vector(sorted_array.active_values());
    cout << "\nIndex of 4: " << sorted_array.search(4) << '\n';
    return 0;
}
